### PR TITLE
Pcam model

### LINKED ISSUE #21: Add PCAM Model & Dataset

### LINKED ISSUE #23: Figure out how to use GPU effectively
I've tried to accelerate training using GPU but I can only get up to about 14% of GPU usage. I think the bottle neck is data loading as the files are loaded dynamically as they are needed ([See PyTorch implementation](https://github.com/pytorch/vision/blob/main/torchvision/datasets/pcam.py), function `__getitem__`). DataLoader allows `num_workers` to be non-zero which in theory should enable dataloading in background but I can't get it to work properly.

Increasing the batch size to 512 helped improve GPU usage to about 24% but at the cost of model performence.

If anyone can figure out how to decrease train time using GPU (by a significant amount) that would be great.

### LINKED ISSUE #25: Confirm that we can use only the center region
On the [official PCAM repo ](https://github.com/basveeling/pcam) it's stated that

> A positive label indicates that the center 32x32px region of a patch contains at least one pixel of tumor tissue. Tumor tissue in the outer region of the patch does not influence the label. This outer region is provided to enable the design of fully-convolutional models that do not use any zero-padding, to ensure consistent behavior when applied to a whole-slide image. This is however not a requirement for the PCam benchmark.

If that's true and we can only train our model on the center 32x32px region then it can speed up our calculations by up to 9 times! Currently we're using [this model](http://cs230.stanford.edu/projects_winter_2019/posters/15813053.pdf) so that would require adjusting.

EDIT: Were we to use 32x32px region I suggest we crop it in our `PCAMX` class so that it's not sent to GPU entirely.

### PR DESCRIPTION
**Pull request by krezelj**
- Add PCAMClassifier
- Add `pcam` option to `ActiveDataset` constructor
- Add PCAMX Class
    - PCAM by default does not implement `classes` and `targets` fields. We rely on these while generating a balanced split. PCAMX is a wrapper class that includes these fields.
    - However if targets were to be loaded in `__init__` all at once it takes about 1 minute of file opening and loading so instead PCAMTargets class is implemented that loads targets lazily as they are needed (currently it doesn't cache them but that might be added).

### CODE REVIEW

**Comment by krezelj** [comment]
Upon further inspection PCAM is still not compatible with `balanced_split=True` and needs more improvements. However that's a separate issue and this PR should only introduce PCAM.

**Comment by krezelj** [comment]
- Added Data Cropping and `PCAMClassifierSmall`
    - Only the center region is needed for actual prediction. Cropping the data allows for smaller and faster models with little to no loss in performence (resolves #25)
- Added `PCAMLazyLoader` that loads the data as needed into a tensor that is stored in RAM.
    - Until now every time an image was needed it had to be loaded from the disk drive. 
    - That operation was a *major* bottleneck in the training process
    - PCAMLazyLoader is essentialy a cache for all of the data. When `pcamx.py` is loaded (e.g. in `data_module.py`) it allocates the necessary space in RAM (a bit less than 4GB for all of the images (8bytes * 3channels * 32w * 32h * 320.000 instances)) for cropped images.
    - PCAMX works with LazyLoader and only loads the data if it is not stored by the LazyLoader
    - The data still has to be loaded at least once and it takes about 2 minutes (with my SSD) to load 50k images. After that the data is stored in RAM and it takes no time to provide it to the model speeding up the training process immensely.
    - Usually most of the data is first loaded when the first query is run. So the more queries are run the more time is saved. In general the more epochs are run the better.
    - Optimised data loading allows GPU to be better utilised increasing its load to about 40% (resolves #23)
    - It is important to note that LazyLoader sacrifices RAM for less computation time. If your machine has less than 8GB of RAM the whole dataset along with the model might take up too much memory.
    - Ideally LazyLoader should be off by default and by default it should not allocate all of the memory at once (**caveat** --appending new data to a list (or dictionary with indices as keys) might ultimately take much more memory than just using a pre-allocated tensor)

Apart from the aforementioned changes this PR introduces some changes throughout the whole project to make it more compatible with PCAM. It should be closed before any new work is started.

**Review by maciejors — APPROVED** [review]
apologies for a late review. code looks great and I think there is no doubt anymore that it works lol

**Comment by maciejors** [comment]
docs look good to me too

**Comment by krezelj** [comment]
Documentation looks good. If you can create a new branch off of this one and continue writing the documentation there I suggest a simple name like `documentation`. No need to create one branch per file.
